readthedocs/notifications/views.py

```diff
@@ -0,0 +1,95 @@
+from django.views.generic import FormView
+from django.contrib import admin, messages
+from django.http import HttpResponseRedirect
+
+from .forms import SendNotificationForm
+
+
+class SendNotificationView(FormView):
+
+    """Form view for sending notifications to users from admin pages
+
+    Accepts the following additional parameters:
+
+    queryset
+        The queryset to use to determine the users to send emails to
+
+    :cvar action_name: Name of the action to pass to the form template,
+                       determines the action to pass back to the admin view
+    :cvar notification_classes: List of :py:class:`Notification` classes to
+                                display in the form
+    """
+
+    form_class = SendNotificationForm
+    template_name = 'notifications/send_notification_form.html'
+    action_name = 'send_email'
+    notification_classes = []
+
+    def get_form_kwargs(self):
+        """Override form kwargs based on input fields
+
+        The admin posts to this view initially, so detect the send button on
+        form post variables. Drop additional fields if we see the send button.
+        """
+        kwargs = super(SendNotificationView, self).get_form_kwargs()
+        kwargs['notification_classes'] = self.notification_classes
+        if 'send' not in self.request.POST:
+            kwargs.pop('data', None)
+            kwargs.pop('files', None)
+        return kwargs
+
+    def get_initial(self):
+        """Add selected ids to initial form data"""
+        initial = super(SendNotificationView, self).get_initial()
+        initial['_selected_action'] = self.request.POST.getlist(
+            admin.ACTION_CHECKBOX_NAME)
+        return initial
+
+    def form_valid(self, form):
+        """If form is valid, send notification to recipients"""
+        count = 0
+        notification_cls = form.cleaned_data['source']
+        for obj in self.get_queryset().all():
+            for recipient in self.get_object_recipients(obj):
+                notification = notification_cls(object=obj,
+                                                request=self.request,
+                                                user=recipient)
+                notification.send()
+                count += 1
+        if count == 0:
+            self.message_user("No recipients to send to", level=messages.ERROR)
+        else:
+            self.message_user("Queued {0} messages".format(count))
+        return HttpResponseRedirect(self.request.get_full_path())
+
+    def get_object_recipients(self, obj):
+        """An iterator for each recipient address for an object
+
+        This allows for non-User querysets to pass back a list of users to send
+        to. By default, assume we're working with :py:class:`User` objects.
+        """
+        yield obj
```

Comment:
That's why I implemented this as an iterator, it should always be treated as an iterable -- perhaps more docs are needed here? The concept here is that this view isn't always going to be tied to `User` objects, it will be used on objects with relations directly and indirectly to `User` objects. It's necessary that the view is tied to the admin queryset for `Project` or whatever, so we need a secondary method to get all of the recipients -- multiple project owners/etc.